Update guidlines on envConfigs and patching to XRs (#549)
crossplane/docs: content/master/concepts/environment-configs.md

description: "Environment Configurations or EnvironmentConfigs are an in-memory datastore used in patching Compositions"
---

A Crossplane EnvironmentConfig is a cluster scoped 

> [ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/) used

Reviewer: ```suggestion
[ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/)-like resource used
```

> Crossplane supports multiple EnvironmentConfigs, each acting as a unique

Author: I rewrote the entire intro now, take a look and let me know what you think.

data store. 

A Composition defines access to one or more EnvironmentConfigs.  

> When Crossplane creates a composite resource, Crossplane creates a unique copy
> of the defined EnvironmentConfigs for that composite resource.  

Reviewer: ```suggestion
When Crossplane reconciles a composite resource, Crossplane merges all the specified EnvironmentConfigs into an in-memory environment.
```

Author: Slight tweak, are you okay with this? 

> When Crossplane creates a composite resource, Crossplane merges all referenced
EnvironmentConfigs and creates a unique in-memory environment
for that composite resource.

Reviewer: 👍

The composite resource can read and write data to their unique 
EnvironmentConfig.

{{<hint "important" >}}

> EnvironmentConfigs are unique to each composite resource.  
> A composite resource can't read data in an EnvironmentConfig written by another
> composite resource.

Reviewer: ```suggestion
The environment is unique to each composite resource.  
A composite resource can't read data in an environment written by another
composite resource.
```

    - item2
```

<!-- vale Google.Headings = NO -->
## Select an EnvironmentConfig
<!-- vale Google.Headings = YES -->

> Select the EnvironmentConfig to use

Reviewer: ```suggestion
Select the EnvironmentConfigs to use
```

for the key {{<hover label="byLabelOptionalDefault" line="14">}}my-second-label-key{{</hover>}}.
If the label
{{<hover label="byLabelOptionalDefault" line="17">}}my-second-label-key{{</hover>}}
exists, Crossplane uses the value from the label instead.

```yaml {label="byLabelOptionalDefault",copy-lines="all"}
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: example-composition
spec:
  environment:
    environmentConfigs:
      - type: Selector
        selector:
          matchLabels:
            - key: my-first-label-key
              type: Value
              value: my-label-value
            - key: my-second-label-key
              type: Value
              value: my-default-value
            - key: my-second-label-key
              type: FromCompositeFieldPath
              valueFromFieldPath: spec.parameters.deploy
              fromFieldPathPolicy: Optional
  resources:
  # Removed for brevity
```

{{<hint "warning" >}}
Crossplane applies values in order. The value of the last key defined always takes precedence.

Defining the default value _after_ the label always overwrites the label
value.
{{< /hint >}}

> 

Reviewer: I couldn't find these info anywhere else, was this intentional?

Author: No, this was an accident because the content was only in master and got deleted when I was copy/pasting between the files and missed this diff. Thanks for catching it!

{{<hover label="maxMatch" line="18">}}matchLabels{{</hover>}} are then merged
into any other environments listed in the 
{{<hover label="maxMatch" line="7">}}environmentConfigs{{</hover>}}.

<!--
TODO: Add Policies
-->

## Patching with EnvironmentConfigs

> When Crossplane creates a composite resource, Crossplane creates a unique copy  
> of the EnvironmentConfig for the composite resource. 

Reviewer: ```suggestion
When Crossplane reconciles a composite resource, Crossplane merges all the specified EnvironmentConfigs into an in-memory environment.
```

Author: I know "reconciles" is the technically correct work but I don't like it, it's too much about the guts of k8s/crossplane.

Would it accurate to say "creates or updates" a composite resource?

Reviewer: Yeah, could be fine

{{< hover label="xrpatch" line="7">}}patches{{</hover>}} object inside of a
Composition's 
{{< hover label="xrpatch" line="5">}}environment{{</hover>}}.

Use the 
{{< hover label="xrpatch" line="5">}}ToCompositeFieldPath{{</hover>}} to copy
data from the EnvironmentConfig to the composite resource.  
Use the 
{{< hover label="xrpatch" line="5">}}FromCompositeFieldPath{{</hover>}} to copy
data from the composite resource to the EnvironmentConfig.

```yaml {label="xrpatch",copy-lines="none"}
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
# Removed for Brevity
spec:
  environment:
  # Removed for Brevity
      patches:
      - type: ToCompositeFieldPath
        fromFieldPath: tags
        toFieldPath: metadata.labels[envTag]
      - type: FromCompositeFieldPath
        fromFieldPath: metadata.name
        toFieldPath: newEnvironmentKey
```

Individual resources can use any data written to the EnvironmentConfig.

##### Patch an individual resource

To patch an individual resource, inside the 
{{<hover label="envpatch" line="16">}}patches{{</hover>}} object of the 
resource, use 
{{<hover label="envpatch" line="17">}}ToEnvironmentFieldPath{{</hover>}} to copy

> data from the resource to the EnvironmentConfig.  

Reviewer: see suggestions in the same section in content/master/concepts/compositions.md
